### ISSUE TITLE
How do we get to the overall metadata

### ISSUE THREAD

**Issue by iherman** [w3c/csvw#145]
This issue has comment threads to issues #105 and #144. The goal is to clarify the full process on how one gets to the overall metadata for a tabular data.

As @gkellogg already described in #144 : 

> In the Model document [Locating Metadata](http://w3c.github.io/csvw/syntax/#locating-metadata) section, the processing order of metadata gives the following precedence:
> 1. User-supplied metadata,
> 2. Embedded metadata,
> 3. Metadata from a Link header,
> 4. File-specific metadata, and
> 5. Directory-specific metadata

and, separately, we have the [specification of the `import` property for a metadata](http://w3c.github.io/csvw/metadata/index.html#importing-metadata). However, there are two possible interpretations of how to get to the final metadata:
1. The metadata files described in the model document are processed in order; the first "hit" counts  and, from that point on, `import` statement in that metadata would determine what other metadata files are (possibly) imported and merged using the semantics of `import`. All other metadata down that priority lane are ignored (I think this is what @gkellogg described in issue #144)
2. _All_ the metadata files listed in the model documents are accessed (if available); each of them are processed separately taking care of the `import` properties (recursively) and then the resulting metadata files are merged according to the same algorithm. Some kind of a "depth-first" merge.

I _think_ that what we discussed at the F2F corresponds to item No. 1 above. But there is a flaw. If there is an embedded metadata (i.e., first row providing the column names) then, according to those rules, the metadata from the link header or the others will _not_ be considered. Similarly if one uses a user-supplied metadata: the first row and all the others will be ignored. While this might still work for the user defined metadata (the user should use an `import`), this is not true for the embedded metadata whose syntax we do not define, i.e., not clear how one would put an `import` statement. Bottomline: that approach does not really work, which leaves us with item No. 2.

And, let us face it, No. 2 is pretty complex. So let me ask this: do we _really_ need that? Wouldn't our earlier approach (i.e., pre-F2F) be better, namely:
1. Remove the `import` property from the metadata vocabulary.
2. The algorithm should be item 2 above, i.e., _all_ metadata listed should be accessed, if available, and then merged using the priority as described and the algorithm currently listed in the metadata document for `import`.

I remember one of the issues listed was that it may require unnecessary HTTP request to find out that, say, a directory-specific metadata does not exist, but I am not convinced that is such a huge issue... Ie, we may want to reconsider our F2F decision.

**Comment by gkellogg** [comment]
My interpretation is slightly different; I don't see a requirement for an explicit `import` (at least, I don't recall the discussion), but my interpretation of steps 1-5 is slightly different:

Step 1 is performed if user explicitly specifies metadata to use. Step 1 is always performed, if the input file is a CSV (or similar), rather than the metadata file itself. Steps 3-5 are performed on a first-found basis. That is, if you retrieve a CSV and the Link header points to a metadata file, that's used. Otherwise, look for -metadata.json and lastly just metadata.json.

So, I would say 2. above without the need for an explicit `import` property, with the provision that only the first found from steps 3-5 is merged in.

Note that processing of the CSV is still complicated by the potential for extracting dialect specifications from 3-5 before even the headers from the CSV can be properly interpreted; it seems we need two passes, first process 1 followed by 3-5 to get dialect information, then using that retrieve embedded metadata (2). Then merge in the order 1-5 to get the metadata used to read rows from the CSV.

**Comment by iherman** [comment]
> On 05 Jan 2015, at 20:55 , Gregg Kellogg notifications@github.com wrote:
> 
> My interpretation is slightly different; I don't see a requirement for an explicit import (at least, I don't recall the discussion), but my interpretation of steps 1-5 is slightly different:
> 
> Step 1 is performed if user explicitly specifies metadata to use. Step 1 is always performed, if the input file is a CSV (or similar), rather than the metadata file itself.

You mean step 2, right:-)

> Steps 3-5 are performed on a first-found basis. That is, if you retrieve a CSV and the Link header points to a metadata file, that's used. Otherwise, look for -metadata.json and lastly just metadata.json.
> 
> So, I would say 2. above without the need for an explicit import property, with the provision that only the first found from steps 3-5 is merged in.

Yes, I see; this makes sense. But the original complaint remains: this is underspecified. And I continue to wonder whether the extra complication of a general `import` is necessary/needed or not. My claim it is not, we can simply take all the metadata files and merge them. This makes the implementation (eg, in the highly async world of Javascript) way easier, because I know in advance the number of files I may potentially handle asynchronously, instead of having some sort of a recursive async structure when handling `import`.

> Note that processing of the CSV is still complicated by the potential for extracting dialect specifications from 3-5 before even the headers from the CSV can be properly interpreted; it seems we need two passes, first process 1 followed by 3-5 to get dialect information, then using that retrieve embedded metadata (2). Then merge in the order 1-5 to get the metadata used to read rows from the CSV.

Hm. I must admit that is not the way I looked upon these dialect information: for me, those were descriptive information only and not prescriptive; i.e., they are used for _describing_ what a parser has done and not defining how a parser should operate. Now that I am reading it I wonder whether that makes sense.

From a practical/implementation point of view this is an extra complication beyond what you describe. Indeed, I, for example, would rely on an external CSV parser, so not only I would have to do those extra passes, but I would have to map the general metadata in terms of CSVW to the way the parser expects the information (let alone the fact that some CSV parsers may be guessing some of those values). Was it really the intention?

We will have to peek into the mind of @JeniT :-)

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by JeniT** [comment]
I've made a new commit (c03166592d5e419936ca403a6d9e00bfe9838485) that includes some [worked-through examples](http://w3c.github.io/csvw/syntax/#examples) of how an annotated tabular data model is created in different circumstances. It doesn't spell out everything, but I'm hopeful this is a start.

**Comment by 6a6d74** [comment]
@JeniT ... looking at the [worked example](http://w3c.github.io/csvw/syntax/#examples) you provide I note that in **5.1.1.2 Using a Metadata File** the `title` properties from the embedded metadata and the metadata file are de-duplicated; e.g. title `GID` appears only once for the first column but it is defined in both locations. (etc.)

```
ID  table   number  cells       name    title                   dc:description
C1  T       1       C1.1, C2.1  GID     GID, Generic Identifier An identifier for the operation on a tree.
```

In the [CSV2RDF mapping](http://w3c.github.io/csvw/publishing-snapshots/FPWD-csv2rdf/Overview.html) I took the approach to add values of `title` from all sources without attempting to de-duplicate. The result of this can be seen in [Example 9](http://w3c.github.io/csvw/publishing-snapshots/FPWD-csv2rdf/Overview.html#map-annotated-tab-examples-2-output) where the `title` properties are used to populate `rdfs:label`-s for the `GID` predicate:

```
:GID a rdf:Property ;
  rdfs:label "GID" , "GID" , "Generic Identifier" ;
  dc:description "An identifier for the operation on a tree." .
```

Do you think that parsers should attempt simple de-duplication of matching strings for `title` properties?

...

Also (& unrelated), when interpreting the header line from the CSV file, I took this to provide _both_ `name` and `title` properties. Your example shows the header line populating only `title`.

**Comment by 6a6d74** [comment]
As an aside, your examples suggest to me that I could re-write the RDF and JSON mapping documents to deal with the _annotated table_ ignoring the the source of those metadata annotations. If this makes sense, I can create a new ISSUE to do so ...

**Comment by iherman** [comment]
> On 08 Jan 2015, at 10:10 , Jeremy Tandy notifications@github.com wrote:
> 
> As an aside, your examples suggest to me that I could re-write the RDF and JSON mapping documents to deal with the annotated table ignoring the the source of those metadata annotations. If this makes sense, I can create a new ISSUE to do so ...

I think it would be very helpful if we could go that way. And also cleaner...

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
Also, note that for testing purposes, we will use RDF isomorphism or SPARQL ASK query, and multiple object values like this that are equivalent don't survive; neither does order, unless they're in a list, which also makes them less useful.

**Comment by gkellogg** [comment]
@JeniT: your examples are useful, but note that by not defining at least one normative way of specifying external metadata we can't test this (or at least a convention to be implemented for testing purposes). I would suggest that one way that MUST be supported is to provide a reference to a metadata file in the standard format.

Also, as you suggest in the examples, all metadata, user-specified or embedded should be converted to the standard representation so that the behavior of merging such information is well defined in terms of the merge algorithm.

**Comment by gkellogg** [comment]
Resolved in PR #169 and on http://www.w3.org/2015/01/28-csvw-minutes.html.
